File: Week_02/G20200343030632/Leetcode_590_632.py
#
# @lc app=leetcode.cn id=590 lang=python3
#
# [590] N叉树的后序遍历
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
class Solution:
    def postorder(self, root: 'Node') -> List[int]:
        return self.postorder_v2(root)
    def postorder_v1(self, root: 'Node') -> List[int]:
        res = []
        def helper(root: 'Node', res: List[int]):
            if root:
                for child in root.children:
                    helper(child, res)
                res.append(root.val)
        helper(root, res)
        return res

    def postorder_v2(self, root: 'Node') -> List[int]:
        res, stack = [], root and [root]
        while stack:
            cur = stack.pop()
            res.append(cur.val)
            stack += cur.children
        return res[::-1]



    
# @lc code=end

File: Week_02/G20200343030632/Leetcode_590_632.java
import java.util.Collections;
import java.util.Deque;
import java.util.LinkedList;


/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> postorder(Node root) {
        // return postorder_v1(root);
        return postorder_v2(root);
    }
    public List<Integer> postorder_v1(Node root) {
        // 递归
        List<Integer> res = new LinkedList<>();
        helper(root, res);
        return res;

    }
    private void helper(Node node, List<Integer> res) {
        if (node != null) {
            for (Node child : node.children) {
                helper(child, res);
            }
            res.add(node.val);
        }
    }
    public List<Integer> postorder_v2(Node root) {
        Deque<Node> stack = new LinkedList<>();
        List<Integer> res = new LinkedList<>();
        if (root == null) {
            return res;
        }
        stack.addFirst(root);
        while (!stack.isEmpty()) {
            Node node = stack.removeFirst();
            res.add(node.val);
            for (Node child : node.children) {
                stack.addFirst(child);
            }
        }
        Collections.reverse(res);
        return res;
    }
}